### ISSUE TITLE
NoWhitespaceBefore, NoWhitespaceAfter, and SeparatorWrap: add support for method reference operator

### ISSUE THREAD

**Issue by eric-milles** [checkstyle/checkstyle#3327]
"::" is separator, NOT a operator - https://docs.oracle.com/javase/specs/jls/se8/html/jls-3.html#jls-3.11

$ cat MethodRef.java

``` java
import java.util.Objects;
import java.util.stream.Stream;

public class MethodRef
{
    {
        Stream.of().filter(Objects :: nonNull); // 2 VIOLATIONS -- want no spaces around '::' operator

        Stream.of().filter(Objects
            ::nonNull); // VIOLATION -- want '::' operator at end of previous line
    }
}
```

$ cat config.xml

``` xml
<!DOCTYPE module PUBLIC
  "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
  "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">
        <module name="NoWhitespaceBefore">
            <message key="ws.preceded" value="Whitespace before ''{0}'' should be removed" />
            <property name="tokens" value="DOUBLE_COLON" />
        </module>
        <module name="NoWhitespaceAfter">
            <message key="ws.followed" value="Whitespace after ''{0}'' should be removed" />
            <property name="tokens" value="DOUBLE_COLON" />
        </module>
        <module name="SeparatorWrap">
            <message key="line.previous" value="Separtor ''{0}'' should be at end of previous line" />
            <property name="tokens" value="DOUBLE_COLON" />
            <property name="option" value="eol" />
        </module>
    </module>
</module>
```

/var/tmp $ java -jar checkstyle-7.0-all.jar -c config.xml MethodRef.java
C:\Users\U0076777>java -jar checkstyle-7.0-all.jar -c config.xml MethodRef.java
com.puppycrawl.tools.checkstyle.api.CheckstyleException: cannot initialize module TreeWalker - Token "DOUBLE_COLON" was
not found in Acceptable tokens list in check com.puppycrawl.tools.checkstyle.checks.whitespace.NoWhitespaceBeforeCheck
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:407)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:138)
        at com.puppycrawl.tools.checkstyle.Main.runCheckstyle(Main.java:385)
        at com.puppycrawl.tools.checkstyle.Main.runCli(Main.java:323)
        at com.puppycrawl.tools.checkstyle.Main.main(Main.java:161)
Caused by: com.puppycrawl.tools.checkstyle.api.CheckstyleException: Token "DOUBLE_COLON" was not found in Acceptable tokens list in check com.puppycrawl.tools.checkstyle.checks.whitespace.NoWhitespaceBeforeCheck
        at com.puppycrawl.tools.checkstyle.TreeWalker.registerCheck(TreeWalker.java:226)
        at com.puppycrawl.tools.checkstyle.TreeWalker.setupChild(TreeWalker.java:163)
        at com.puppycrawl.tools.checkstyle.api.AutomaticBean.configure(AutomaticBean.java:138)
        at com.puppycrawl.tools.checkstyle.Checker.setupChild(Checker.java:402)
        ... 4 more
Checkstyle ends with 1 errors.

---

Want to enforce no whitespace around the new '::' operator.  Also, need DOUBLE_COLON added to OperatorWrap check to be able to enforce the wrapping policy.
##

**Comment by rnveach** [comment]
`DOUBLE_COLON` is not a valid token currently. Token needed to add is `METHOD_REF`.

**Comment by rnveach** [comment]
Closing as issue was split into 3 separate issues
#3949, #3950, #3951
